src/Orleans/Messaging/SqlMembershipTable.cs

```diff
@@ -21,489 +21,229 @@ MIT License
+            database = RelationalStorageUtilities.CreateGenericStorageInstance(WellKnownRelationalInvariants.SqlServer, config.DataConnectionString);
+
             return TaskDone.Done;
         }
 
+
         public TimeSpan MaxStaleness
         {
             get { return maxStaleness; }
         }
 
+
         public bool IsUpdatable
         {
             get { return true; }
         }
 
+
         public IList<Uri> GetGateways()
         {
-            var result = new List<Uri>();
-
-            // NOTE: This method is only called from client, so blocking operations should be ok.
-
-            using (var conn = new SqlConnection(connectionString))
+            logger.Verbose3("SqlMembershipTable.GetGateways called.");
+            try
             {
-                conn.Open();
-
-                var cmdStr = String.Format(CLIENT_READ_ROW_SELECT, (Int32)SiloStatus.Active);
-                var command = new SqlCommand(cmdStr);
-                command.Parameters.Add(new SqlParameter { ParameterName = "@deploymentid", DbType = DbType.String, Value = deploymentId });
-                command.Connection = conn;
-
-                // Note: This does blocking read, so must not be called from grain code.
-                using (var results = command.ExecuteReader())
-                {
-                    while (results.Read())
-                    {
-                        int proxyPort = 0;
-                        if (!results.GetSqlInt32(1).IsNull)
-                            proxyPort = results.GetInt32(1);
-                        result.Add(new IPEndPoint(IPAddress.Parse(results.GetString(0)), proxyPort).ToGatewayUri());
-                    }
-                }
+                //TODO: Refactor this to async.
+                return database.ActiveGatewaysAsync(deploymentId).Result;
             }
-            return result;
-        }
-
-        async Task<MembershipTableData> IMembershipTable.ReadRow(SiloAddress key)
-        {
-            using (var conn = new SqlConnection(connectionString))
+            catch(Exception ex)
             {
-                conn.Open();
-
-                var command = new SqlCommand(READ_ROW_SELECT, conn);
-                command.Parameters.Add(new SqlParameter { ParameterName = "@deploymentid", DbType = DbType.String, Value = deploymentId });
-                command.Parameters.Add(new SqlParameter { ParameterName = "@address", DbType = DbType.String, Value = key.Endpoint.Address.ToString() });
-                command.Parameters.Add(new SqlParameter { ParameterName = "@port", DbType = DbType.Int32, Value = key.Endpoint.Port });
-                command.Parameters.Add(new SqlParameter { ParameterName = "@generation", DbType = DbType.Int32, Value = key.Generation });
-
-                return await ProcessResults(conn, command);
+                logger.Verbose("SqlMembershipTable.Gateways failed {0}", ex);
```

Comment:
I don't know if ``.Verbose`` is the correct level to log membership errors. The tasks are awaited in lower SQL level and the tasks are basically pass-through here, so not all problems are catched here, I think.